### PR TITLE
Better Airdrop production [Superseded]

### PR DESCRIPTION
**Pull request by penev92**
This is a direct successor to #7061, with the remaining requested changes and some refactoring by me.
This adds support for multiple-actor delivery to ProductionAirdrop and moves it to Mods.Common.
This does (should) not change the current behavior at all. Only lays the foundations for the upcoming Bulk production.

P.S.: Also now the delivery aircraft will wait at the drop location while the "produced" actor exits before flying off (in case the exit is blocked and the actor can't exit right away).

### CODE REVIEW

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 1]
```diff
@@ -321,7 +322,7 @@
     <Compile Include="Traits\Player\PlaceBeacon.cs" />
     <Compile Include="Traits\Player\PlaceBuilding.cs" />
     <Compile Include="Traits\Player\PlayerStatistics.cs" />
-    <Compile Include="Traits\Player\ProductionQueue.cs" />
+    <Compile Include="Traits\Buildings\ProductionQueue.cs" />
```
Comment (pchote) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: This has nothing inherently to do with buildings.  In most mods it lives on the Player actor.

Comment (pchote) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: All our production traits would be better in their own Production directory.

File: OpenRA.Mods.Common/Traits/Buildings/Production.cs [hunk 2]
```diff
@@ -93,27 +95,30 @@ public void DoProduction(Actor self, ActorInfo producee, ExitInfo exitinfo, stri
...
-		public virtual bool Produce(Actor self, ActorInfo producee, string raceVariant)
+		public virtual bool Produce(Actor self, IEnumerable<ActorInfo> actorsToProduce, string raceVariant)
 		{
 			if (Reservable.IsReserved(self))
 				return false;
 
-			// pick a spawn/exit point pair
-			var exit = self.Info.Traits.WithInterface<ExitInfo>().Shuffle(self.World.SharedRandom)
-				.FirstOrDefault(e => CanUseExit(self, producee, e));
+			var produced = self;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/Production.cs]: `null` would be a more appropriate default for this.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/Production.cs]: Hang on, this is never even assigned to anything else.

File: OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+				{
+					new CenterPositionInit(w.Map.CenterOfCell(startPos) + new WVec(WRange.Zero, WRange.Zero, altitude)),
+					new OwnerInit(owner),
+					new FacingInit(64)
+				});
+
+				Action func = null;
+				func = () =>
+				{
+					if (actorInfos.Count == 0)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs]: `if (!actorInfos.Any())`.

File: OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+				var altitude = self.World.Map.Rules.Actors[deliveringActorType].Traits.Get<PlaneInfo>().CruiseAltitude;
+				var a = w.CreateActor(deliveringActorType, new TypeDictionary
+				{
+					new CenterPositionInit(w.Map.CenterOfCell(startPos) + new WVec(WRange.Zero, WRange.Zero, altitude)),
+					new OwnerInit(owner),
+					new FacingInit(64)
+				});
+
+				Action func = null;
+				func = () =>
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs]: This monster anonymous function should be split out into at least one proper method.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj (hunk 1 above)
Comment (penev92) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: What do you mean? I see only TD use `ProductionQueue` and that is on the factories, not the players.
Yeah that is a good idea.

File: OpenRA.Mods.Common/Traits/Buildings/Production.cs (hunk 2 above)
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Buildings/Production.cs]: Hm, not sure what @WolfGaming had in mind here. I'll look into it.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj (hunk 1 above)
Comment (pchote) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: It holds all of the base logic for the `ClassicProductionQueue` used in the other mods.

**Comment by penev92** [comment]
Updated. Created a `Mods.Common\Traits\Production` folder for some of the production traits.
Tested in D2k and TD and saw no regressions. Even with the "Instant build" speed and spamming units.

**Comment by WolfGaming** [comment]
Thanks for taking this over @penev92

File: OpenRA.Mods.Common/Traits/Production/Production.cs [hunk 5]
```diff
@@ -93,27 +95,32 @@ public void DoProduction(Actor self, ActorInfo producee, ExitInfo exitinfo, stri
...
-		public virtual bool Produce(Actor self, ActorInfo producee, string raceVariant)
+		public virtual bool Produce(Actor self, IEnumerable<ActorInfo> actorsToProduce, string raceVariant)
 		{
 			if (Reservable.IsReserved(self))
 				return false;
 
-			// pick a spawn/exit point pair
-			var exit = self.Info.Traits.WithInterface<ExitInfo>().Shuffle(self.World.SharedRandom)
-				.FirstOrDefault(e => CanUseExit(self, producee, e));
+		    var allProduced = true;
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Production/Production.cs]: Spaces detected.

File: OpenRA.Mods.Common/Traits/Production/Production.cs [hunk 6]
```diff
@@ -93,27 +95,32 @@ public void DoProduction(Actor self, ActorInfo producee, ExitInfo exitinfo, stri
...
 			{
-				DoProduction(self, producee, exit, raceVariant);
-				return true;
+				// pick a spawn/exit point pair
+				var exit = self.Info.Traits.WithInterface<ExitInfo>().Shuffle(self.World.SharedRandom)
+					.FirstOrDefault(e => CanUseExit(self, actorInfo, e));
+			    if (exit != null)
+			        DoProduction(self, actorInfo, exit, raceVariant);
+			    else
+			        allProduced = false;
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Production/Production.cs]: Again, we really need that StyleCop+ rule, which sadly depends on https://github.com/shuruev/StyleCopPlus/pull/5.

File: OpenRA.Mods.Common/Traits/Production/ProductionAirdrop.cs [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+				deliveringActor.QueueActivity(new Land(Target.FromActor(self)));
+				deliveringActor.QueueActivity(new CallFunc(() => MakeDelivery(actorsToProduce.ToList(), deliveringActor, w)));
+			});
+
+			return true;
+		}
+
+		private void MakeDelivery(ICollection<ActorInfo> actorInfos, Actor deliveringActor, World world)
+		{
+		    if (!actorInfos.Any())
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Production/ProductionAirdrop.cs]: This is all spaces again. Try setting your editor on tabs per project.

**Comment by penev92** [comment]
Updated btw.

**Comment by Mailaender** [comment]
> P.S.: Also now the delivery aircraft will wait at the drop location while the "produced" actor exits before flying off (in case the exit is blocked and the actor can't exit right away).

This does not work in Dune 2000 when you order a carryall from the starport. It will never unload it and keep sitting on the landing bay:

image

**Comment by penev92** [comment]
Hm, wonder when that came back. OK, should be easily fixable:
https://github.com/OpenRA/OpenRA/pull/7061#issuecomment-65406319

**Comment by penev92** [comment]
Fixed and rebased.

File: OpenRA.Mods.Common/Traits/Production/ProductionAirdrop.cs [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+							nbm.OnNotifyBlockingMove(blocker, self);
+
+					WaitAndRetry(deliveringActor, world, actorInfos);
+				}
+				else
+				{
+					actorInfos.Remove(actorInfo);
+
+					world.AddFrameEndTask(w =>
+						{
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Production/ProductionAirdrop.cs]: Wrong indentation here.

**Comment by obrakmann** [comment]
> P.S.: Also now the delivery aircraft will wait at the drop location while the "produced" actor exits before flying off (in case the exit is blocked and the actor can't exit right away).

That looks very glitchy, especially with multiple outstanding orders (the delivery aircraft just stack on top of one another on the airfield).

Also the notification on completed production is missing.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 9]
```diff
@@ -203,7 +203,7 @@ public bool AircraftCanEnter(Actor a)
 
 		public bool IsLeavingCell(CPos location, SubCell subCell = SubCell.Any) { return false; } // TODO: Handle landing
 		public SubCell GetValidSubCell(SubCell preferred) { return SubCell.Invalid; }
-		public SubCell GetAvailableSubCell(CPos a, SubCell preferredSubCell = SubCell.Any, Actor ignoreActor = null, bool checkTransientActors = true) { return SubCell.Invalid; } // Does not use any subcell
+		public SubCell GetAvailableSubCell(CPos a, SubCell preferredSubCell = SubCell.Any, Actor ignoreActor = null, bool checkTransientActors = true) { return SubCell.Any; } // Does not use any subcell
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: What's the story behind this change?

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: This also invalidates the comment.

**Comment by pchote** [comment]
Production should block if an aircraft is already on the pad, like we do if a regular exit is blocked.

File: OpenRA.Mods.Common/Traits/Production/ProductionAirdrop.cs [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+using System.Drawing;
+using System.Linq;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Deliver the unit in production via skylift.")]
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/Production/ProductionAirdrop.cs]: airlift

Comment (penev92) [inline OpenRA.Mods.Common/Traits/Production/ProductionAirdrop.cs]: To be fair, "spacelift" 50% of the time :D

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 9 above)
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: See @Mailaender's comment above and mine after it.

Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: I assume that's caused by [this check](https://github.com/OpenRA/OpenRA/pull/7468/files#diff-86d6fc75cf06769f960cd874106931f7R125) then.

File: OpenRA.Mods.Cnc/Traits/Buildings/ProductionAirdrop.cs [hunk 11]
```diff
@@ -1,90 +0,0 @@
...
-				a.QueueActivity(new CallFunc(() =>
-				{
-					if (!self.IsInWorld || self.IsDead)
-						return;
-
-					foreach (var cargo in self.TraitsImplementing<INotifyDelivery>())
-						cargo.Delivered(self);
-
-					self.World.AddFrameEndTask(ww => DoProduction(self, producee, exit, raceVariant));
-					Sound.PlayNotification(self.World.Map.Rules, self.Owner, "Speech", info.ReadyAudio, self.Owner.Country.Race);
```
Comment (obrakmann) [inline OpenRA.Mods.Cnc/Traits/Buildings/ProductionAirdrop.cs]: This line is missing in your new implementation.

**Comment by obrakmann** [comment]
Looks good overall, but needs some things fixed:

- [ ] missing notification
- [ ] prevent multiple aircraft from landing at the same time
- [ ] fix the introduced hack for delivering aircraft

File: OpenRA.Mods.Common/Traits/Production/ProductionAirdrop.cs (hunk 10 above)
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Production/ProductionAirdrop.cs]: `ProductionAirdrop` not `ProductionSpacedrop`.

**Comment by pchote** [comment]
It looks like this has stalled for now, so closing.   Adding it to the future milestone so that you (or someone else) can resurrect this in the future since we are going to need this at some point.

**Comment by pchote** [comment]
I tested the original C&C to see what it does when the airfield is blocked:

1. If the airfield is blocked when an actor finished construction it will block until it is cleared (like we do for non-airfields in OpenRA).
2. If a plane was already enroute before it becomes blocked it will fly off the map and refund the cost of the cargo.

**Comment by penev92** [comment]
Does the original C&C airstrip wait for the delivery to be finished (or maybe even for the delivery plane to go away) before starting to build the next unit on the queue?

**Comment by penev92** [comment]
Resurrected in #7786.

**Pull request closed without merging**